coding tracker for review

jakubmikolajewski opened the pull request.

jakubmikolajewski committed:
coding tracker for review

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/CodingController.cs:
```diff
@@ -0,0 +1,279 @@
+﻿using Dapper;
+
+namespace CodingTrackerLibrary
```
Remember as per the [code-conventions](https://thecsharpacademy.com/article/58/code-conventions): 
From C# 10 we are able to ditch the namespace block and save one indentation level, which optimizes the use of space.

![code-convention-6](https://github.com/user-attachments/assets/f813d72d-94f4-41d6-aec8-f37371eab50a)

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/CodingController.cs:
```diff
@@ -0,0 +1,279 @@
+            string command = $"SELECT * FROM 'Coding sessions';";
+
+            DatabaseActions current = new DatabaseActions();
+            using (current.Connection)
```
I am not exactly sure on your thinking here. But the initialisation of a DatabaseActions object is redundant and makes the code harder to read. Might even stop what you are trying to achieve with the using statement.

I would have just kept is simple with:
`using var connection = new SqlLiteConnection(connectionString);`

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/CodingController.cs:
```diff
@@ -0,0 +1,279 @@
+                    {
+                        while (reader.Read())
+                        {
+                            CodingSession session = new CodingSession(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetInt32(3));
```
The beauty of Dapper is that you no longer have to use a Reader object and all this horrible boilerplate code. This can be simplified to:
```
var sessions = current.Connection.Query<CodingSession>(command).ToList();
TableVisualizationEngine.selectAllSessions.AddRange(sessions);
```

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/CodingGoals.cs:
```diff
@@ -0,0 +1,19 @@
+        public string GoalName {get;set;}
+        public int DesiredHours {get;set;}
+        public DateTime ReachGoalBy {get;set;}
+        public CodingGoals(long goalId, string goalName, int desiredHours, string reachGoalBy)
```
Remember your Spacing [code-conventions](https://thecsharpacademy.com/article/58/code-conventions):

Have one empty line between methods.

![image](https://github.com/user-attachments/assets/427f15da-01cb-400a-a6d4-b345f7bb1f22)

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/CodingSession.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    public class CodingSession
+    {
+        private long _id;
```
A wonderful thing about modern C# is that you no longer need these backing fields in the current way you are using them (there may be instances where you do, but looking at this class, you don't!).

Your property declaration on line 25 could simply be 
`public long Id { get; }`

All the magic happens in the background and you don't clutter your class up.

chrisjamiecarter commented on codingTracker.jakubmikolajewski/CodingTrackerLibrary/DatabaseActions.cs:
```diff
@@ -0,0 +1,60 @@
+                }   
+            }
+        }
+
```
RE: Spacing. No empty lines are necessary after the last closing bracket.

chrisjamiecarter reviewed (changes requested):
Hey @jakubmikolajewski 👋 ,

Congratulations on another project submission! 🎉 It’s great to see your progress.

- Perfect project submission structure.
- Very well written README.md, it is great to see you communicate what you have developed, your challenges and lessons learned.

I have added some comments to your code to help you improve. These do not need action, just review and see if you can apply next project. Also I have only commented on the first instance of an issue I spot. The same issue may be present elsewhere in your code.

**Bugs**
I have found a bug that does need action though!

![image](https://github.com/user-attachments/assets/5b4ebeb5-23b1-477c-b5c2-77c2c21dff05)

As you can see I am entering a start time and end time (NOTE: it asks me for a start date twice which is confusing!)

Then when I view all sessions I get an exception:

![image](https://github.com/user-attachments/assets/b1371dc5-52e7-463c-af11-a45c1d811c4b)


Feel free to reach out if you have any questions or if you'd like any help.

Once fixed, commit and push your changes. This will update this Pull Request and notify me to continue my review.

Thanks
@chrisjamiecarter

jakubmikolajewski committed:
Changes after review

Thank you for taking the time to review this project! :)
Refactored:
-corrected DateTime parsing
-now asking for date only once (not sure what I was thinking before)
-corrected goal to only show remaining time if goal has not been reached yet
-some minor stylistic changes
Thank you for the other comments too, I have already started the Flashcards project and have been trying to apply these principles better.

chrisjamiecarter reviewed (approved):
Well done @jakubmikolajewski on another completed project 🎉
